internal/subscriber/subscriberimpl_test.go

```diff
@@ -0,0 +1,301 @@
+			wantErr:          true,
+		},
+	}
+
+	for _, test := range tests {
+		messages := make([]messagebroker.ReceivedMessage, 0, test.maxNumOfMessages)
+		for index, msg := range test.expected {
+			pubSub := &metrov1.PubsubMessage{Data: []byte(msg)}
+			data, _ := proto.Marshal(pubSub)
+			msgProto := messagebroker.ReceivedMessage{
+				Data:      data,
+				Topic:     Topic,
+				Partition: Partition,
+			}
+			msgProto.MessageID = strconv.Itoa(index)
+			messages = append(messages, msgProto)
+		}
+
+		cs.EXPECT().ReceiveMessages(ctx, messagebroker.GetMessagesFromTopicRequest{NumOfMessages: test.maxNumOfMessages, TimeoutMs: 1000}).Return(
+			&messagebroker.GetMessagesFromTopicResponse{
+				Messages: messages,
+			}, test.err,
+		)
+
+		pullRequest := &PullRequest{ctx: ctx, MaxNumOfMessages: test.maxNumOfMessages}
+		responseChan := make(chan *metrov1.PullResponse, 10)
+		errChan := make(chan error, 10)
+		subImpl.Pull(ctx, pullRequest, responseChan, errChan)
+
+		ticker := time.NewTimer(30 * time.Second)
+		select {
+		case resp := <-responseChan:
+			assert.Equal(t, len(test.expected), len(resp.ReceivedMessages))
+			for index, msg := range resp.ReceivedMessages {
+				assert.Equal(t, test.expected[index], string(msg.Message.Data))
+			}
+		case err := <-errChan:
+			assert.Equal(t, test.wantErr, err != nil)
+		case <-ticker.C:
+			assert.FailNow(t, "Test case timed out")
+		}
+	}
+}
+
+func TestBasicImplementation_Acknowledge(t *testing.T) {
+	ctx, cs, subImpl := setup(t)
+	testInputs := []struct {
+		message            string
+		offset             int32
+		messageID          string
+		canConsumeMore     bool
+		maxCommittedOffset int32
+	}{
+		{
+			message:            "there",
+			offset:             2,
+			messageID:          "1",
+			canConsumeMore:     false,
+			maxCommittedOffset: 0,
+		},
+		{
+			message:            "Hello",
+			offset:             1,
+			messageID:          "2",
+			canConsumeMore:     true,
+			maxCommittedOffset: 2,
+		},
+		{
+			message:            "Razor",
+			offset:             3,
+			messageID:          "3",
+			canConsumeMore:     true,
+			maxCommittedOffset: 3,
+		},
+	}
+
+	messages := make([]messagebroker.ReceivedMessage, 0, 3)
+	for _, input := range testInputs {
+		data, _ := proto.Marshal(&metrov1.PubsubMessage{Data: []byte(input.message)})
+		msgProto := messagebroker.ReceivedMessage{
+			Data:      data,
+			Topic:     Topic,
+			Partition: Partition,
+			Offset:    input.offset,
+		}
+		msgProto.MessageID = input.messageID
+		messages = append(messages, msgProto)
+	}
+	cs.EXPECT().ReceiveMessages(ctx, gomock.Any()).Return(
+		&messagebroker.GetMessagesFromTopicResponse{
+			Messages: messages,
+		}, nil,
+	)
+
+	pullRequest := &PullRequest{ctx: ctx, MaxNumOfMessages: int32(len(testInputs))}
+	responseChan := make(chan *metrov1.PullResponse, 1)
+	errChan := make(chan error, 10)
+	subImpl.Pull(ctx, pullRequest, responseChan, errChan)
+
+	ticker := time.NewTimer(30 * time.Second)
```

Comment:
On second thought let's restrict this to 2 seconds. Since this pattern repeats multiple times it can add up to a significant amount of wait time.